src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,310 @@
+          sparqlExpression::copyExpressionResult(
+              sparqlExpression::ExpressionResult{
+                  resultTable.back()[alias.outCol_]});
+    } else {
+      for (const auto& substitution : substitutions) {
+        const auto& occurrences =
+            get<std::vector<GroupBy::ParentAndChildIndex>>(
+                substitution.occurrences_);
+        substituteGroupVariable(
+            occurrences, resultTable.back()[substitution.resultColumnIndex_]);
+      }
+
+      // Substitute in the results of all aggregates contained in the
+      // expression of the current alias, if `info` is non-empty.
+      // Substitute in the results of all aggregates of `info`.
+      std::vector<std::unique_ptr<sparqlExpression::SparqlExpression>>
+          originalChildren;
+      originalChildren.reserve(info.size());
+      for (auto& aggregate : info) {
+        // Substitute the resulting vector as a literal
+        auto newExpression = std::make_unique<sparqlExpression::IdExpression>(
+            calculateAggregateResult(aggregate.aggregateDataIndex_));
+
+        AD_CONTRACT_CHECK(aggregate.parentAndIndex_.has_value());
+        auto parentAndIndex = aggregate.parentAndIndex_.value();
+        originalChildren.push_back(std::move(
+            parentAndIndex.parent_->children()[parentAndIndex.nThChild_]));
+        parentAndIndex.parent_->replaceChild(parentAndIndex.nThChild_,
+                                             std::move(newExpression));
+      }
+
+      // Evaluate top-level alias expression
+      sparqlExpression::ExpressionResult expressionResult =
+          alias.expr_.getPimpl()->evaluate(&evaluationContext);
+
+      // Restore original children
+      for (size_t i = 0; i < info.size(); ++i) {
+        auto& aggregate = info.at(i);
+        auto parentAndIndex = aggregate.parentAndIndex_.value();
+        parentAndIndex.parent_->replaceChild(parentAndIndex.nThChild_,
+                                             std::move(originalChildren.at(i)));
+      }
+
+      // Copy the result so that future aliases may reuse it
+      evaluationContext._previousResultsFromSameGroup.at(alias.outCol_) =
+          sparqlExpression::copyExpressionResult(expressionResult);
+
+      // Extract values
+      resultTable.back()[alias.outCol_] = std::visit(
+          [this]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) -> Id {
+            constexpr static bool isStrongId = std::is_same_v<T, Id>;
+            AD_CONTRACT_CHECK(sparqlExpression::isConstantResult<T>);
+            if constexpr (isStrongId) {
+              return singleResult;
+            } else if constexpr (sparqlExpression::isConstantResult<T>) {
+              return sparqlExpression::detail::constantExpressionResultToId(
+                  AD_FWD(singleResult), localVocab_);
+            } else {
+              // This should never happen since aggregates always return
```

Comment:
Are you telling me, that this is not 90% duplication of something that we already have in the HashMap optimization, that can be unified to your needs?